Source/Tools/LinuxEmulation/LinuxSyscalls/Seccomp/BPFEmitter.cpp

```diff
@@ -0,0 +1,397 @@
+// SPDX-License-Identifier: MIT
+/*
+$info$
+tags: LinuxSyscalls|syscalls-shared
+$end_info$
+*/
+
+#include "LinuxSyscalls/Seccomp/BPFEmitter.h"
+#include "LinuxSyscalls/Seccomp/SeccompEmulator.h"
+
+#include <linux/bpf_common.h>
+#include <linux/filter.h>
+#include <linux/seccomp.h>
+
+namespace FEX::HLE {
+template<bool CalculateSize>
+uint64_t BPFEmitter::HandleLoad(uint32_t BPFIP, const sock_filter* Inst) {
+  [[maybe_unused]] size_t OpSize {};
+  switch (BPF_SIZE(Inst->code)) {
+  case BPF_H:
+  case BPF_B:
+  case 0x18: /* BPF_DW */ RETURN_ERROR(-EINVAL);
+  case BPF_W: break;
+  }
+
+  const auto DestReg = BPF_CLASS(Inst->code) == BPF_LD ? REG_A : REG_X;
+
+  switch (BPF_MODE(Inst->code)) {
+  case BPF_IMM: {
+    auto Const = ConstPool.try_emplace(Inst->k, ARMEmitter::ForwardLabel {});
+    EMIT_INST(ldr(DestReg, &Const.first->second););
+    break;
+  }
+  case BPF_ABS: {
+    // ABS has some restrictions
+    // - Must be 4-byte aligned
+    // - Must be less than the size of seccomp_data
+    const auto Offset = Inst->k;
+    if (Offset & 0b11) {
+      ///< Wasn't 4-byte aligned.
+      RETURN_ERROR(-EINVAL);
+    }
+
+    if (Offset >= sizeof(seccomp_data)) {
+      ///< Tried accessing outside of seccomp_data.
+      RETURN_ERROR(-EINVAL);
+    }
+
+    EMIT_INST(ldr(DestReg, REG_SECCOMP_DATA, Offset););
+    break;
+  }
+  case BPF_MEM:
+    if (Inst->k >= 16) {
+      ///< Larger than scratch space size.
+      RETURN_ERROR(-EINVAL);
+    }
+
+    EMIT_INST(ldr(DestReg, REG_SECCOMP_DATA, offsetof(WorkingBuffer, ScratchMemory[Inst->k])););
```

Comment:
`EMIT_INST(ldr(..););` isn't too pretty, and EMIT_INST requires that there's exactly 1 instruction emitted anyway for correct size calculation. Could we insert the trailing `;` in EMIT_INST itself and then `s/;);/);/g` throughout this file?

Reply:
Macros that don't end on `;` at use site combine poorly with clang-format, see #3878.

Reply:
I'm talking about the other `;`

Reply:
Ah, the thought crossed my mind but then I read "trailing" and thought you meant the latter. Apologies :)

Reply:
Moved the semi-colon in to the define so we don't have the double semi-colon thing.